dvc/stage/run.py

```diff
@@ -13,12 +16,37 @@
 logger = logging.getLogger(__name__)
 
 
+def _get_shell_version(executable):
+    name = os.path.basename(executable).lower()
+    out = subprocess.run(
+        [executable, "-c", f"echo -n ${name.upper()}_VERSION"],
+        stdout=subprocess.PIPE,
+    ).stdout.decode("utf-8")
+
+    return re.match(r"\d+\.\d+\.?\d?", out).group()
+
+
 def _make_cmd(executable, cmd):
     if executable is None:
         return cmd
-    opts = {"zsh": ["--no-rcs"], "bash": ["--noprofile", "--norc"]}
+
+    def get_opts(name):
+        version = _get_shell_version(executable)
```

Comment:
So we are calling this for every invocation no matter the shell. This is bad, we should only do this for fish, really.

Reply:
I agree. I think even for fish it's a bit much. Hence, my unhappiness.

Reply:
It's even invoked twice currently. Once, when checking for fish, once in `_make_cmd`. This could be easily refactored to one call, though.

Reply:
Well, if there is no other way of doing that then running a command - i think that is fine. The overhead is pretty minimal and should only affect fish users, but it makes it usable on fish, so that's a big win and is worth the cost.

Reply:
Btw, how long ago was the first version supporting `--no-config` released? And also what is the upgrading process for typical fish users? I'm just wondering whether fish users always use the latest version (e.g. maybe there is an aggressive updating or even auto-updating policy) and we can always assume that `--no-config` works? Maybe would also check if old fish versions error out if you pass `--no-config` to them, just in case it is odd like that and won't complain.

Reply:
Well, fish does not auto-update. `--no-config` was merged and release about a year ago. For example Debian bullseye (which I consider to be still around and use 😄 ) still have fish 3.1.

I have currently no fish pre 3.3 around, but
```shell
❯ fish --abc
fish: unrecognized option '--abc'
```
does not get my hopes up.

Reply:
@clamydo Gotcha. Then let's check for it for fish.